OCPBUGS-5497: MCDRebootError alarm disappears after 15 minutes

djoshy opened the pull request:
<!--
If this is a bug fix, make sure your description includes "Fixes: #xxxx", or
"Closes: #xxxx"

Please provide the following information:
-->

**- What I did**
On reboot, MCD now goes into a loop that updates the metric counter on every timeout(5 minutes) & attempts to reboot again. The MCDRebootError alert rule sees the metric counter increase over time and stays latched past 15 minutes(..forever essentially). 

**- How to verify it**
Follow reproduction rules from the [bug](https://issues.redhat.com/browse/OCPBUGS-5497) to block a reboot. Alert will appear 5 minutes after the first failed reboot attempt and should sustain until the node is rebooted. 

**- Description for the changelog**
OCPBUGS-5497: daemon: added reboot reconciliation; counter for total reboot attempts

<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

yuqi-zhang commented on pkg/daemon/update.go:
```diff
@@ -2155,23 +2155,26 @@ func (dn *Daemon) reboot(rationale string) error {
-	if err := rebootCmd.Run(); err != nil {
-		dn.logSystem("failed to run reboot: %v", err)
+	rebootCmd := rebootCommand(rationale)
+	for {
```
I agree here, the for loop seems a bit dangerous, since the MCD will no longer be able to reconcile any changes or react to updates that potentially clear this problem (although that is unlikely so it's not the end of the world).

Since we only sleep 5 minutes now, what happens if you remove this for loop entirely and just keep the contents inside? (or enqueue a node sync in 5 minutes instead of sleeping here)? I would think the MCD would try to update again and fail on this error, and so long as we do that within the 15 minutes the alarm would potentially still stay?

djoshy replied:
Thanks for the feedback Jerry! Here are the changes I made

-removed the loop, left the 5 minute timeout in
-Added an immediate return if the command fails,  it didn't make sense for us to wait in this scenario. this still increments the counter(at a much faster pace), so it should still set the alert
-The other path is still a 5 minute timeout via a sleep
-Added/Changed some additional logging explaining both paths

the other thing to note is that once a reboot recovery happens(by this, I mean we have a fail, and then the reboot somehow works, not sure if this is even possible/realistic) - the alert will take 15 minutes to go away

sinnykumari commented on pkg/daemon/update.go:
```diff
@@ -2155,24 +2155,26 @@ func (dn *Daemon) reboot(rationale string) error {
 	}
-
-	// wait to be killed via SIGTERM from the kubelet shutting down
+	// wait 5 minutes to see if MCD will be killed via SIGTERM from the kubelet shutting down
```
I think we no longer need this timer now. With PR https://github.com/openshift/machine-config-operator/pull/2509 , we no longer explicitly stopping kubelet. Everything is getting handled by `systemctl reboot`  . So ideally if that fails, we should increase the counter and return error.

djoshy replied:
Just updated the PR with our latest findings! (:

djoshy commented:
So after some more discussion with Sinny here is what I did/found:
-I initially assumed rebootCmd.run() never returns on a successful reboot, this is incorrect
-Removed the timeout, so MCO only attempts to reconcile the reboot on error from `rebootCmd.Run()`. This means that if `rebootCmd.Run()` returns successfully, we will assume reboot is imminent & not throw an error
-Changed the prometheus rule to absolute check vs a rate/increase check. This will clear the alert immediately on a successful reboot

yuqi-zhang commented on pkg/daemon/update.go:
```diff
@@ -2155,24 +2155,20 @@ func (dn *Daemon) reboot(rationale string) error {
-	mcdRebootErr.Inc()
-	return fmt.Errorf("reboot failed; this error should be unreachable, something is seriously wrong")
+	// if we're here, reboot went through successfully
+	dn.logSystem("reboot successful")
```
So in theory, it is possible for the MCO to run another sync after this returns? i.e. If the reboot is slow enough to sigterm'ing the MCO, is it possible for us to have started another update?

Looking at our other use the the timeout we do have in `RunFirstbootCompleteMachineconfig`, looks like that doesn't do quite the same thing, but we use the timeout to stop the MCD from doing another sync. We could do that for extra extra safety if we wanted to 

I'm fine with us not worrying about that path in this PR as well, just a thought

sinnykumari replied:
I too have some doubt here on whether to keep some sleep duration. But, I am not convinced (or have enough data) on what purpose it will serve by adding some sleep before reboot. Also, if we want to keep some sleep duration, on what parameter we determine how much we should keep.

@cgwalters Perhaps, you may have some more insight on does it makes sense to keep `time.Sleep(defaultRebootTimeout)` considering we no longer explicitly stop kubelet service with RR https://github.com/openshift/machine-config-operator/pull/2509

cgwalters replied:
Agree we can remove the sleep here; I think that was only necessary before graceful node shutdown.

djoshy replied:
sounds good, thanks for the feedback! (:

yuqi-zhang commented on install/0000_90_machine-config-operator_01_prometheus-rules.yaml:
```diff
@@ -67,7 +67,7 @@ spec:
         - alert: MCDRebootError
           expr: |
-            max by(namespace, node, pod) (increase(mcd_reboots_failed_total[15m])) > 0
+            mcd_reboots_failed_total > 0
```
I'm a bit of an alert newbie, when you say `This will clear the alert immediately on a successful reboot`, what is processing that? I see we do a `Inc()` on the alert which is what I assume is keeping this active, but what clears it if we reboot after say, 5 failures? Is it purely time based?

sinnykumari replied:
counter metrics get reset only when server gets restarted. In our case, successful reboot will cause MCD metrics running on the node set to 0 . https://prometheus.io/docs/instrumenting/writing_clientlibs/#counter .

djoshy replied:
yep, as sinny said the counters get reset to 0, so that should clear the error. I think increase would've worked just as well, but seems a bit overkill since we don't need to look for transients/jumps but rather anything non zero

yuqi-zhang reviewed (commented):
Overall makes sense to me! 2 questions below

sinnykumari commented on install/0000_90_machine-config-operator_01_prometheus-rules.yaml:
```diff
@@ -67,7 +67,7 @@ spec:
         - alert: MCDRebootError
           expr: |
-            max by(namespace, node, pod) (increase(mcd_reboots_failed_total[15m])) > 0
+            mcd_reboots_failed_total > 0
```
with this change, I would be interested in adding a `for: 5min` to fire alert only when reboot attempt is failing after multiple attempts over 5min. @jan--f Does this makes sense?

djoshy replied:
oh good thought, if we want to give it some slack to recover, this makes sense. we could also get a similar effect by changing the expression `max by(namespace, node, pod) (increase(mcd_reboots_failed_total[5m])) > 0`

jan--f replied:
Yep that would work.

djoshy replied:
Updated with a `for: 5min`  expression

djoshy commented:
Taking this out of draft as we all are in agreement with the current changes

sinnykumari commented:
This looks good to me now! Thanks David for keeping all the patience and going through all the changes required :)
/approve

@yuqi-zhang any final thoughts?

Putting hold for qe-approved label
/hold

yuqi-zhang reviewed (commented):
/lgtm

Seems fine to me! If there are any additional weird behaviour resulting from the timeout removal, I think it wouldn't be fatal regardless, and we can always retroactively monitor the tests to see if anything crops up

djoshy commented:
/test e2e-aws-ovn-upgrade

sergiordlr commented:
Verified using IPI on AWS.

Steps:

1. Break the reboot process
```
$ mount -o remount,rw /usr
$ mv /usr/bin/systemd-run /usr/bin/systemd-run2

```
2. Creat any MC. For example, this one:

```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name: test-file
spec:
  config:
    ignition:
      version: 3.1.0
    storage:
      files:
      - contents:
          source: data:text/plain;charset=utf;base64,c2VydmVyIGZvby5leGFtcGxlLm5ldCBtYXhkZWxheSAwLjQgb2ZmbGluZQpzZXJ2ZXIgYmFyLmV4YW1wbGUubmV0IG1heGRlbGF5IDAuNCBvZmZsaW5lCnNlcnZlciBiYXouZXhhbXBsZS5uZXQgbWF4ZGVsYXkgMC40IG9mZmxpbmUK
        filesystem: root
        mode: 0644
        path: /etc/test
```


The node is degraded

```
$ oc get mcp
NAME     CONFIG                                             UPDATED   UPDATING   DEGRADED   MACHINECOUNT   READYMACHINECOUNT   UPDATEDMACHINECOUNT   DEGRADEDMACHINECOUNT   AGE
master   rendered-master-1ed414611751fc2d99874185ab65fe79   True      False      False      3              3                   3                     0                      55m
worker   rendered-worker-93e0d654b9f6b8106d5d1e3f178ab219   False     True       True       3              0                   0                     1                      55m

```

The reboot command fails and the MCD tries to execute again every few minutes. `mcd_reboots_failed_total` metric is  increased in every retry.

```
I0203 14:08:43.147473    2251 update.go:1897] Removing SIGTERM protection
I0203 14:08:43.147495    2251 update.go:1867] initiating reboot: Node will reboot into config rendered-worker-92ed488450cea709233430372d43c173
I0203 14:08:43.149257    2251 update.go:1867] failed to run reboot: exec: "systemd-run": executable file not found in $PATH

```


After 5 minutes aprox, a MCDRebootError alert is triggered:

```
$ curl -s -k -H "Authorization: Bearer $(oc -n openshift-monitoring create token prometheus-k8s)" https://$(oc get route -n openshift-monitoring alertmanager-main -o jsonpath={.spec.host})/api/v1/alerts | jq 
....
    {
      "labels": {
        "alertname": "MCDRebootError",
        "container": "oauth-proxy",
        "endpoint": "metrics",
        "instance": "10.0.162.79:9001",
        "job": "machine-config-daemon",
        "namespace": "openshift-machine-config-operator",
        "node": "ip-10-0-162-79.us-east-2.compute.internal",
        "openshift_io_alert_source": "platform",
        "pod": "machine-config-daemon-9ptsf",
        "prometheus": "openshift-monitoring/k8s",
        "service": "machine-config-daemon",
        "severity": "critical"
      },
      "annotations": {
        "message": "Reboot failed on ip-10-0-162-79.us-east-2.compute.internal , update may be blocked. For more details:  oc logs -f -n openshift-machine-config-operator machine-config-daemon-9ptsf -c machine-config-daemon "
      },
...
```

After 20 minutes the alert is not removed.

3. Fix the reboot process

```
$ oc debug node/....
.....
$ mount -o remount,rw /usr
$ mv /usr/bin/systemd-run2 /usr/bin/systemd-run
```

The MCD logs show that the reboot could be executed successfully

```
I0203 14:10:48.290519    2251 update.go:1867] initiating reboot: Node will reboot into config rendered-worker-92ed488450cea709233430372d43c173
I0203 14:10:48.301061    2251 update.go:1867] reboot successful
I0203 14:10:48.322463    2251 daemon.go:1019] Shutting down MachineConfigDaemon
I0203 14:10:48.322610    2251 metrics.go:98] Metrics listener successfully stopped

```

The `mcd_reboots_failed_total`  metric  is reset to 0.

The alert is removed.

We add the qe-approved label

/label qe-approved
